gk/fib.c

```diff
@@ -0,0 +1,1326 @@
+					ip_prefix);
+				return -1;
+			}
+
+			/*
+			 * TODO Add concurrency control to
+			 * reset the Ethernet cache entry.
+			 */
+			eth_cache->stale = false;
+			memset(&eth_cache->eth_hdr, 0,
+				sizeof(eth_cache->eth_hdr));
+			eth_cache->ref_cnt = 0;
+		}
+
+		break;
+	}
+
+	case GK_DROP:
+		break;
+
+	case GK_FWD_NEIGHBOR_FRONT_NET:
+		/* FALLTHROUGH */
+	case GK_FWD_NEIGHBOR_BACK_NET:
+		/* FALLTHROUGH */
+	default:
+		return -1;
+		break;
+	}
+
+	/*
+	 * If the updated FIB entry doesn't have an action GK_FWD_GRANTOR,
+	 * we don't need to flush the flow tables at all.
+	 */
+	if (ip_prefix_fib->action == GK_FWD_GRANTOR) {
+
+		rte_atomic16_init(&ip_prefix_fib->num_updated_instances);
+
+		/* Send the FIB entry to the GK mailboxes. */
+		for (i = 0; i < gk_conf->num_lcores; i++) {
+			notify_flow_flush(ip_prefix_fib,
+				&gk_conf->instances[i]);
+		}
+
+		/*
+		 * Wait until all the GK instances
+		 * finish to flush their flow tables.
+		 */
+		while (rte_atomic16_read(&ip_prefix_fib->num_updated_instances)
+			!= gk_conf->num_lcores);
+	}
+
+	/* Reset the fields of the deleted FIB entry. */
+	ip_prefix_fib->action = GK_FIB_MAX;
+	ip_prefix_fib->ref_cnt = 0;
+	rte_atomic16_init(&ip_prefix_fib->num_updated_instances);
+	memset(&ip_prefix_fib->u, 0, sizeof(ip_prefix_fib->u));
+
+	RTE_LOG(NOTICE, GATEKEEPER,
+		"gk: delete a fib entry [ip prefix = %s]\n", ip_prefix);
+
+	return 0;
+}
+
+/*
+ * Initialize a gateway FIB entry.
+ * @gateway the gateway address informaiton.
+ * @ip_prefix the IP prefix,
+ * for which the gateway is responsible.
+ */
+static struct gk_fib *
+init_gateway_fib(const char *ip_prefix, enum gk_fib_action action,
+	const char *gateway, struct gk_config *gk_conf)
+{
+	int ret;
+	int fib_id = -1;
+	int ip_prefix_len;
+	struct ipaddr gw_addr;
+	struct ipaddr ip_prefix_addr;
+	struct ether_cache *eth_cache;
+	struct gk_fib *gw_fib = NULL;
+	struct gk_fib *neigh_fib = NULL;
+	struct gk_fib *ip_prefix_fib = NULL;
+	struct gk_lpm *ltbl = &gk_conf->lpm_tbl;
+	struct gatekeeper_if *iface;
+
+	/* Initialize the fib entry for the gateway. */
+	ret = convert_str_to_ip(gateway, &gw_addr);
+	if (ret < 0)
+		return NULL;
+
+	ip_prefix_len = parse_ip_prefix(ip_prefix, &ip_prefix_addr);
+	if (ip_prefix_len < 0)
+		return NULL;
+
+	if (gw_addr.proto != ip_prefix_addr.proto) {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to initialize a fib entry for gateway, since the gateway and its responsible grantors have different IP versions.\n");
+		return NULL;
+	}
+
+	if (action == GK_FWD_GATEWAY_FRONT_NET)
+		iface = &gk_conf->net->front;
+	else if (action == GK_FWD_GATEWAY_BACK_NET)
+		iface = &gk_conf->net->back;
+	else {
+		RTE_LOG(ERR, GATEKEEPER,
+			"gk: failed to initialize a fib entry for gateway, since it has invalid action %d.\n",
+			action);
+		return NULL;
+	}
+
+	/* Find the FIB entry for the gateway IP address. */
+	if (gw_addr.proto == ETHER_TYPE_IPv4 &&
+			ipv4_if_configured(iface)) {
```

Comment:
For this case (and the IPv6 case below), you check to make sure `ipv*_if_configured()` is true. Why are these tests not also done in the `init_grantor_fib()` function? Maybe I am missing the point.